### DISCUSSION TITLE
ModuleNotFoundError: Error when trying to create a Python Data Loader

### DISCUSSION THREAD

**Discussion by alessandrofajr** [observablehq/framework#1720] [Q&A] [answered]
I am trying to create a Data Loader using Python, but I keep encountering the following error: `ModuleNotFoundError: No module named 'pandas'`.

Here is the full traceback:
```
load /data/data.csv → [missing] Traceback (most recent call last):
  File "C:\Users\observableframework\src\data\data.csv.py", line 1, in <module>
    import pandas as pd
ModuleNotFoundError: No module named 'pandas'
error in 636ms: loader exited with code 1
```

### Details:

In the file `src\data\data.csv.py`, here is my code:

```
import pandas as pd
import numpy as np

import sys

# Simple random DataFrame
df = pd.DataFrame({
    'A': np.random.randint(1, 10, 5),
    'B': np.random.rand(5)             
})

df.to_csv(sys.stdout, index=False)
```
### My Setup:
OS: Windows 11
Python: 3.12.2
Node: 20.11.0
Observable Framework: 1.11.0

### Steps I followed:

- I created a new project using the Observable Framework with `npx "@observablehq/framework@latest" create` in an empty folder.
- Then, I created a Python virtual environment in the same folder using `python -m virtualenv .venv`.
- I activated the virtual environment with `.venv\Scripts\activate`.
- I added pandas and numpy to a `requirements.txt` file located at the same folder as `package.json` files and ran `pip install -r requirements.txt`

I tried calling my Data Loader from a .md file like this:

```js
const myData = FileAttachment("data/data.csv").csv({typed: true});
```

The error occurs for any Python package that isn't part of the standard library. I’ve tried creating new projects with clean installs, but nothing seems to work.

If I run something like `python src\data\data.csv.py`, the code runs just fine.

**Comment by mythmon** [comment]
~~Are you activating the the virtual environment? I don't know what that looks like on Windows, but on other platforms you would run `source .venv/bin/activate`. You need to do this in each terminal where you are going to run `npm run build` so that it knows to use the virtualenv.~~

Apologies, I see that you already said you did this.

**Comment by mbostock** [comment]
Did you start the preview server within the virtual environment as described in [Data loaders: Execution](https://observablehq.com/framework/data-loaders#execution)?

**Reply by alessandrofajr** [reply]
Yes, with the virtualenv activated, I ran `npm run dev`

**Comment by mythmon** [comment, marked as answer]
Another thing to check out, by default Framework uses the binary name `python3` to run the Python interpreter, not `python`. If you run `python3 src\data\data.csv.py`, does the code work?

If you want to use just `python`, you could set the [`interpreters`](https://observablehq.com/framework/config#interpreters) option to that:

```js
{
  interpreters: {
    ".py": ["python"],
  }
}
```

**Reply by alessandrofajr** [reply]
This did the trick! Thank you
